src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -62,6 +66,171 @@ public static Document Current
+        /// Purge unused Elements from the model.
+        /// </summary>
+        /// <param name="deepPurge">Similar to clicking the purge button multiple times in Revit.</param>
+        /// <returns>Purged element ids</returns>
+        public List<int> PurgeUnused(bool deepPurge = false)
+        {
+            //The internal GUID of the purge Performance Adviser Rule 
+            const string purgePerformanceAdviserRuleGuid = "e8c63650-70b7-435a-9010-ec97660c1bda";
+
+            IList<PerformanceAdviserRuleId> performanceAdviserRules = PerformanceAdviser.GetPerformanceAdviser().GetAllRuleIds();
+            List<PerformanceAdviserRuleId> performanceAdviserRuleId = performanceAdviserRules.Where(x => x.Guid.ToString() == purgePerformanceAdviserRuleGuid).ToList();
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            TransactionManager.Instance.EnsureInTransaction(this.InternalDocument);
+            
+            purgedElementIds.AddRange(PurgeElements(performanceAdviserRuleId, deepPurge));
+            purgedElementIds.AddRange(PurgeMaterials());
+
+            TransactionManager.Instance.TransactionTaskDone();
+
+            if (purgedElementIds.Count == 0)
+                throw new InvalidOperationException(Properties.Resources.NoElementsToPurge);
+
+            return purgedElementIds.Select(x => x.IntegerValue).ToList();
+        }
+
+        private List<ElementId> PurgeMaterials()
+        {
+            List<Autodesk.Revit.DB.ElementId> materials = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(Autodesk.Revit.DB.Material))
+                                                                                                           .ToElementIds()
+                                                                                                           .ToList();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument).WhereElementIsElementType()
+                                                                                                              .ToElements()
+                                                                                                              .ToList();
+            var nonPurgedElements = new List<Autodesk.Revit.DB.ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                HostObjAttributes hostObj = elementTypes[i] as HostObjAttributes;
+                if (hostObj != null)
+                {
+                    var compundStructure = hostObj.GetCompoundStructure();
+                    if (compundStructure == null)
+                        continue;
+
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        ElementId elementId = compundStructure.GetMaterialId(j);
+                        if (materials.Contains(elementId))
```

Comment:
probably pointless to keep checking on every iteration.

a more efficient method would be to keep track of ids you encounter, then do a list difference at the end between all materials and the ones encountered, which will give you the ones that aren't used and hence can be purged.

Reply:
Maybe not understanding this correctly, but if I do a new list with the encountered ids wouldn't I still need to do the check to see if I already added the id to the list?

Reply:
not if you 
- use a `HashSet` instead of `List`
- use a `Distinct()` before removing the ids/deleting/whatever needs doing